File: src/ves_module.zig
const std = @import("std");
const fs = std.fs;
const diag = @import("diagnostic.zig");
const zon = std.zon;

const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;
const Io = std.Io;
const Span = @import("location.zig").Span;

const getConfigPath = @import("Config.zig").getConfigPath;

const VESTI_DUMMY_DIR = @import("vesti-info").VESTI_DUMMY_DIR;

pub const VestiModule = struct {
    name: []const u8,
    version: ?[]const u8,
    exports: []const struct {
        name: []const u8,
        location: ?[]const u8 = null,
    },
};

pub fn downloadModule(
    allocator: Allocator,
    diagnostic: *diag.Diagnostic,
    mod_name: []const u8,
    import_file_loc: ?Span,
) !void {
    var mod_dir_path = try ArrayList(u8).initCapacity(allocator, 30);
    defer mod_dir_path.deinit(allocator);

    const config_path = try getConfigPath(allocator);
    defer allocator.free(config_path);

    try mod_dir_path.print(allocator, "{s}/{s}", .{ config_path, mod_name });

    var mod_data_path = try ArrayList(u8).initCapacity(
        allocator,
        mod_dir_path.items.len + 15,
    );
    defer mod_data_path.deinit(allocator);
    try mod_data_path.print(allocator, "{s}/vesti.zon", .{mod_dir_path.items});

    var mod_zon_file = fs.cwd().openFile(mod_data_path.items, .{}) catch {
        const io_diag = try diag.IODiagnostic.init(
            allocator,
            import_file_loc,
            "cannot open file {s}",
            .{
                mod_data_path.items,
            },
        );
        diagnostic.initDiagInner(.{ .IOError = io_diag });
        return error.FailedGetModule;
    };
    defer mod_zon_file.close();

    var buf: [1024]u8 = undefined;
    var mod_zon_file_reader = mod_zon_file.reader(&buf);

    // what kind of such simple config file has 4MB size?
    // TODO: if 0.16.0 is finalized, replace this code with the following
    //
    //config_zon_reader.interface.allocRemainingAlignedSentinel(
    //  allocator,
    //  .limited(4 * 1024 * 1024),
    //  .of(u8),
    //  0,
    //)
    const context = blk: {
        const reader = &mod_zon_file_reader.interface;
        var tmp = Io.Writer.Allocating.init(allocator);
        defer tmp.deinit();

        var remaining: Io.Limit = .limited(4 * 1024 * 1024);
        while (remaining.nonzero()) {
            const n = Io.Reader.stream(reader, &tmp.writer, remaining) catch |err| switch (err) {
                error.EndOfStream => break :blk try tmp.toOwnedSliceSentinel(0),
                error.WriteFailed => break :blk error.OutOfMemory,
                error.ReadFailed => break :blk error.ReadFailed,
            };
            remaining = remaining.subtract(n).?;
        }
        break :blk error.StreamTooLong;
    } catch {
        const io_diag = try diag.IODiagnostic.init(
            allocator,
            null,
            "cannot read context from {s}",
            .{config_path},
        );
        diagnostic.initDiagInner(.{ .IOError = io_diag });
        return error.FailedOpenConfig;
    };
    defer allocator.free(context);

    const ves_module = try zon.parse.fromSlice(
        VestiModule,
        allocator,
        context,
        null,
        .{},
    );
    defer zon.parse.free(allocator, ves_module);

    for (ves_module.exports) |@"export"| {
        var mod_filename = try ArrayList(u8).initCapacity(
            allocator,
            @"export".name.len + mod_dir_path.items.len,
        );
        defer mod_filename.deinit(allocator);
        try mod_filename.print(
            allocator,
            "{s}/{s}",
            .{ mod_dir_path.items, @"export".name },
        );

        const location = @"export".location orelse VESTI_DUMMY_DIR;
        var into_copy_filename = try ArrayList(u8).initCapacity(
            allocator,
            @"export".name.len + location.len,
        );
        defer into_copy_filename.deinit(allocator);
        try into_copy_filename.print(
            allocator,
            "{s}/{s}",
            .{ location, @"export".name },
        );

        fs.cwd().copyFile(
            mod_filename.items,
            fs.cwd(),
            into_copy_filename.items,
            .{},
        ) catch {
            const io_diag = try diag.IODiagnostic.init(
                allocator,
                import_file_loc,
                "cannot copy from {s} into {s}",
                .{
                    mod_filename.items,
                    into_copy_filename.items,
                },
            );
            diagnostic.initDiagInner(.{ .IOError = io_diag });
            return error.FailedGetModule;
        };
    }
}

File: src/ves_module.rs
use std::fs;
use std::path::PathBuf;

use serde::Deserialize;

use crate::config::get_config_path;
use crate::diagnostic::{Diagnostic, DiagnosticInner, IoDiagnostic};
use crate::location::Span;
use crate::vesti_info::VESTI_DUMMY_DIR;

#[derive(Debug)]
pub enum VesModuleError {
    FailedGetModule,
    FailedOpenConfig,
}

#[derive(Debug, Deserialize)]
pub struct VestiExport {
    pub name: String,
    #[serde(default)]
    pub location: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct VestiModule {
    pub name: String,
    #[serde(default)]
    pub version: Option<String>,
    pub exports: Vec<VestiExport>,
}

pub fn download_module(
    diagnostic: &mut Diagnostic<'_>,
    mod_name: &str,
    import_file_loc: Option<Span>,
) -> Result<(), VesModuleError> {
    let config_path = get_config_path();
    let mod_dir_path: PathBuf = config_path.join(mod_name);
    let mod_data_path = mod_dir_path.join("vesti.ron");

    let context = match fs::read_to_string(&mod_data_path) {
        Ok(c) => c,
        Err(_) => {
            let io_diag = IoDiagnostic::new(
                import_file_loc,
                format!("cannot open file {}", mod_data_path.display()),
            );
            diagnostic.init_diag_inner(DiagnosticInner::IoError(io_diag));
            return Err(VesModuleError::FailedGetModule);
        }
    };

    let ves_module: VestiModule = match ron::from_str(&context) {
        Ok(m) => m,
        Err(_) => {
            let io_diag = IoDiagnostic::new(
                None,
                format!("cannot read context from {}", mod_data_path.display()),
            );
            diagnostic.init_diag_inner(DiagnosticInner::IoError(io_diag));
            return Err(VesModuleError::FailedOpenConfig);
        }
    };
    // `version` is part of the manifest but unused at copy time
    let _ = &ves_module.name;
    let _ = &ves_module.version;

    for export in &ves_module.exports {
        let mod_filename = mod_dir_path.join(&export.name);
        let location = export.location.as_deref().unwrap_or(VESTI_DUMMY_DIR);
        let into_copy_filename = PathBuf::from(location).join(&export.name);

        if fs::copy(&mod_filename, &into_copy_filename).is_err() {
            let io_diag = IoDiagnostic::new(
                import_file_loc,
                format!(
                    "cannot copy from {} into {}",
                    mod_filename.display(),
                    into_copy_filename.display()
                ),
            );
            diagnostic.init_diag_inner(DiagnosticInner::IoError(io_diag));
            return Err(VesModuleError::FailedGetModule);
        }
    }

    Ok(())
}
